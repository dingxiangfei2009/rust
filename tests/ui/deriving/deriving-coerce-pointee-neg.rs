//@ proc-macro: malicious-macro.rs
#![feature(derive_coerce_pointee, arbitrary_self_types)]

extern crate core;
extern crate malicious_macro;

use std::marker::CoercePointee;

#[derive(CoercePointee)]
//~^ ERROR: `CoercePointee` can only be derived on `struct`s with `#[repr(transparent)]`
//~| NOTE: in this expansion of #[derive(CoercePointee)]
enum NotStruct<'a, T: ?Sized> {
    Variant(&'a T),
}

#[derive(CoercePointee)]
//~^ ERROR: `CoercePointee` can only be derived on `struct`s with at least one field
//~| NOTE: in this expansion of #[derive(CoercePointee)]
#[repr(transparent)]
struct NoField<#[pointee] T: ?Sized> {}
//~^ ERROR: type parameter `T` is never used
//~| NOTE: unused type parameter
//~| HELP: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`

#[derive(CoercePointee)]
//~^ ERROR: `CoercePointee` can only be derived on `struct`s with at least one field
//~| NOTE: in this expansion of #[derive(CoercePointee)]
#[repr(transparent)]
struct NoFieldUnit<#[pointee] T: ?Sized>();
//~^ ERROR: type parameter `T` is never used
//~| NOTE: unused type parameter
//~| HELP: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`

#[derive(CoercePointee)]
//~^ ERROR: `CoercePointee` can only be derived on `struct`s that are generic over at least one type
//~| NOTE: in this expansion of #[derive(CoercePointee)]
#[repr(transparent)]
struct NoGeneric<'a>(&'a u8);

#[derive(CoercePointee)]
//~^ ERROR: exactly one generic type parameter must be marked as `#[pointee]` to derive `CoercePointee` traits
//~| NOTE: in this expansion of #[derive(CoercePointee)]
#[repr(transparent)]
struct AmbiguousPointee<'a, T1: ?Sized, T2: ?Sized> {
    a: (&'a T1, &'a T2),
}

#[derive(CoercePointee)]
#[repr(transparent)]
struct TooManyPointees<'a, #[pointee] A: ?Sized, #[pointee] B: ?Sized>((&'a A, &'a B));
//~^ ERROR: only one type parameter can be marked as `#[pointee]` when deriving `CoercePointee` traits
//~| NOTE: here another type parameter is marked as `#[pointee]`

#[derive(CoercePointee)]
struct NotTransparent<'a, #[pointee] T: ?Sized> {
    //~^ ERROR: `derive(CoercePointee)` is only applicable to `struct` with `repr(transparent)` layout
    ptr: &'a T,
}

#[derive(CoercePointee)]
#[repr(transparent)]
struct NoMaybeSized<'a, #[pointee] T> {
    //~^ ERROR: `derive(CoercePointee)` requires `T` to be marked `?Sized`
    ptr: &'a T,
}

#[derive(CoercePointee)]
#[repr(transparent)]
struct PointeeOnField<'a, #[pointee] T: ?Sized> {
    #[pointee]
    //~^ ERROR: the `#[pointee]` attribute may only be used on generic parameters
    ptr: &'a T,
}

#[derive(CoercePointee)]
#[repr(transparent)]
struct PointeeInTypeConstBlock<
    'a,
    T: ?Sized = [u32; const {
                    struct UhOh<#[pointee] T>(T);
                    //~^ ERROR: the `#[pointee]` attribute may only be used on generic parameters
                    10
                }],
> {
    ptr: &'a T,
}

#[derive(CoercePointee)]
#[repr(transparent)]
struct PointeeInConstConstBlock<
    'a,
    T: ?Sized,
    const V: u32 = {
        struct UhOh<#[pointee] T>(T);
        //~^ ERROR: the `#[pointee]` attribute may only be used on generic parameters
        10
    },
> {
    ptr: &'a T,
}

#[derive(CoercePointee)]
#[repr(transparent)]
struct PointeeInAnotherTypeConstBlock<'a, #[pointee] T: ?Sized> {
    ptr: PointeeInConstConstBlock<
        'a,
        T,
        {
            struct UhOh<#[pointee] T>(T);
            //~^ ERROR: the `#[pointee]` attribute may only be used on generic parameters
            0
        },
    >,
}

// However, reordering attributes should work nevertheless.
#[repr(transparent)]
#[derive(CoercePointee)]
struct ThisIsAPossibleCoercePointee<'a, #[pointee] T: ?Sized> {
    ptr: &'a T,
}

// Also, these paths to Sized should work
#[derive(CoercePointee)]
#[repr(transparent)]
struct StdSized<'a, #[pointee] T: ?std::marker::Sized> {
    ptr: &'a T,
}
#[derive(CoercePointee)]
#[repr(transparent)]
struct CoreSized<'a, #[pointee] T: ?core::marker::Sized> {
    ptr: &'a T,
}
#[derive(CoercePointee)]
#[repr(transparent)]
struct GlobalStdSized<'a, #[pointee] T: ?::std::marker::Sized> {
    ptr: &'a T,
}
#[derive(CoercePointee)]
#[repr(transparent)]
struct GlobalCoreSized<'a, #[pointee] T: ?::core::marker::Sized> {
    ptr: &'a T,
}

#[derive(CoercePointee)]
#[malicious_macro::norepr]
#[repr(transparent)]
struct TryToWipeRepr<'a, #[pointee] T: ?Sized> {
    //~^ ERROR: `derive(CoercePointee)` is only applicable to `struct` with `repr(transparent)` layout [E0802]
    ptr: &'a T,
}

#[repr(transparent)]
#[derive(CoercePointee)]
struct RcWithId<T: ?Sized> {
    inner: std::rc::Rc<(i32, Box<T>)>,
    //~^ ERROR: `Box<T>` cannot be coerced to an unsized value [E0802]
    //~| NOTE: `derive(CoercePointee)` demands that `Box<T>` can be coerced to an unsized type
    //~| HELP: `derive(CoercePointee)` requires exactly one copy of `#[pointee]` type at the end of the `struct` definition, without any further pointer or reference indirection
    //~| ERROR: `Box<T>` cannot be coerced to an unsized value [E0802]
    //~| NOTE: `derive(CoercePointee)` demands that `Box<T>` can be coerced to an unsized type
    //~| HELP: `derive(CoercePointee)` requires exactly one copy of `#[pointee]` type at the end of the `struct` definition, without any further pointer or reference indirection
    //~| NOTE: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
}

#[repr(transparent)]
#[derive(CoercePointee)]
struct MoreThanOneField<T: ?Sized> {
    //~^ ERROR: transparent struct needs at most one field with non-trivial size or alignment, but has 2 [E0690]
    //~| NOTE: needs at most one field with non-trivial size or alignment, but has 2
    inner1: Box<T>,
    //~^ ERROR: `derive(CoercePointee)` only admits exactly one data field, to which `dyn` methods shall be dispatched [E0802]
    //~| ERROR: `derive(CoercePointee)` only admits exactly one data field, on which unsize coercion shall be performed [E0802]
    //~| NOTE: this field has non-zero size or requires alignment
    inner2: Box<T>,
    //~^ NOTE: this field has non-zero size or requires alignment
}

struct NotCoercePointeeData<T: ?Sized>(T);

#[repr(transparent)]
#[derive(CoercePointee)]
struct UsingNonCoercePointeeData<T: ?Sized>(NotCoercePointeeData<T>);
//~^ ERROR: `NotCoercePointeeData<T>` cannot be coerced to an unsized type [E0802]
//~| NOTE: `derive(CoercePointee)` demands that `NotCoercePointeeData<T>` can be coerced to an unsized type
//~| HELP: the standard pointers such as `Arc`, `Rc`, `Box`, and other types with `derive(CoercePointee)` can be coerced to their corresponding unsized types
//~| ERROR: `NotCoercePointeeData<T>` cannot be coerced to an unsized type, to which `dyn` methods can be dispatched [E0802]
//~| NOTE: `derive(CoercePointee)` demands that `dyn` methods can be dispatched when `NotCoercePointeeData<T>` can be coerced to an unsized type
//~| HELP: `dyn` methods can be dispatched to the standard pointers such as `Arc`, `Rc`, `Box`, and other types with `derive(CoercePointee)`


fn main() {}
