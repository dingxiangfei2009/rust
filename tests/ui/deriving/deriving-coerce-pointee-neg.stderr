error[E0802]: `CoercePointee` can only be derived on `struct`s with `#[repr(transparent)]`
  --> $DIR/deriving-coerce-pointee-neg.rs:10:10
   |
LL | #[derive(CoercePointee)]
   |          ^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `CoercePointee` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0802]: `CoercePointee` can only be derived on `struct`s that are generic over at least one type
  --> $DIR/deriving-coerce-pointee-neg.rs:30:10
   |
LL | #[derive(CoercePointee)]
   |          ^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `CoercePointee` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0802]: exactly one generic type parameter must be marked as `#[pointee]` to derive `CoercePointee` traits
  --> $DIR/deriving-coerce-pointee-neg.rs:35:10
   |
LL | #[derive(CoercePointee)]
   |          ^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `CoercePointee` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0802]: only one type parameter can be marked as `#[pointee]` when deriving `CoercePointee` traits
  --> $DIR/deriving-coerce-pointee-neg.rs:44:39
   |
LL | struct TooManyPointees<'a, #[pointee] A: ?Sized, #[pointee] B: ?Sized>((&'a A, &'a B));
   |                                       ^                     - here another type parameter is marked as `#[pointee]`

error: the `#[pointee]` attribute may only be used on generic parameters
  --> $DIR/deriving-coerce-pointee-neg.rs:63:5
   |
LL |     #[pointee]
   |     ^^^^^^^^^^

error: the `#[pointee]` attribute may only be used on generic parameters
  --> $DIR/deriving-coerce-pointee-neg.rs:73:33
   |
LL |                     struct UhOh<#[pointee] T>(T);
   |                                 ^^^^^^^^^^

error: the `#[pointee]` attribute may only be used on generic parameters
  --> $DIR/deriving-coerce-pointee-neg.rs:87:21
   |
LL |         struct UhOh<#[pointee] T>(T);
   |                     ^^^^^^^^^^

error: the `#[pointee]` attribute may only be used on generic parameters
  --> $DIR/deriving-coerce-pointee-neg.rs:102:25
   |
LL |             struct UhOh<#[pointee] T>(T);
   |                         ^^^^^^^^^^

error[E0392]: lifetime parameter `'a` is never used
  --> $DIR/deriving-coerce-pointee-neg.rs:18:16
   |
LL | struct NoField<'a, #[pointee] T: ?Sized> {}
   |                ^^ unused lifetime parameter
   |
   = help: consider removing `'a`, referring to it in a field, or using a marker such as `PhantomData`

error[E0392]: type parameter `T` is never used
  --> $DIR/deriving-coerce-pointee-neg.rs:18:31
   |
LL | struct NoField<'a, #[pointee] T: ?Sized> {}
   |                               ^ unused type parameter
   |
   = help: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`

error[E0802]: `derive(CoercePointee)` is only applicable to `struct` with `repr(transparent)` layout
  --> $DIR/deriving-coerce-pointee-neg.rs:48:1
   |
LL | struct NotTransparent<'a, #[pointee] T: ?Sized> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0802]: `derive(CoercePointee)` requires the `#[pointee]` to be `?Sized`
  --> $DIR/deriving-coerce-pointee-neg.rs:55:1
   |
LL | struct NoMaybeSized<'a, #[pointee] T> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0802]: `derive(CoercePointee)` is only applicable to `struct` with `repr(transparent)` layout
  --> $DIR/deriving-coerce-pointee-neg.rs:141:1
   |
LL | struct TryToWipeRepr<'a, #[pointee] T: ?Sized> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0802]: `derive(CoercePointee)` is derived multiple times
  --> $DIR/deriving-coerce-pointee-neg.rs:146:10
   |
LL | #[derive(CoercePointee, CoercePointee)]
   |          ^^^^^^^^^^^^^  ^^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `CoercePointee` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0802]: `derive(CoercePointee)` is derived multiple times
  --> $DIR/deriving-coerce-pointee-neg.rs:146:10
   |
LL | #[derive(CoercePointee, CoercePointee)]
   |          ^^^^^^^^^^^^^  ^^^^^^^^^^^^^
   |
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
   = note: this error originates in the derive macro `CoercePointee` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Box<T>: Unsize<Box<T {coerced}>>` is not satisfied
  --> $DIR/deriving-coerce-pointee-neg.rs:155:10
   |
LL | #[derive(CoercePointee)]
   |          ^^^^^^^^^^^^^ the trait `Unsize<Box<T {coerced}>>` is not implemented for `Box<T>`
   |
   = note: all implementations of `Unsize` are provided automatically by the compiler, see <https://doc.rust-lang.org/stable/std/marker/trait.Unsize.html> for more information
   = note: this error originates in the derive macro `CoercePointee` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0802]: `CoercePointee` can only be derived on `struct`s with at least one field with a `pointee` type
  --> $DIR/deriving-coerce-pointee-neg.rs:18:1
   |
LL | struct NoField<'a, #[pointee] T: ?Sized> {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0802]: `CoercePointee` can only be derived on `struct`s with at least one field with a `pointee` type
  --> $DIR/deriving-coerce-pointee-neg.rs:25:1
   |
LL | struct NoFieldUnit<'a, #[pointee] T: ?Sized>();
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `Box<T>: Unsize<Box<T {coerced}>>` is not satisfied
  --> $DIR/deriving-coerce-pointee-neg.rs:155:10
   |
LL | #[derive(CoercePointee)]
   |          ^^^^^^^^^^^^^ the trait `Unsize<Box<T {coerced}>>` is not implemented for `Box<T>`
   |
   = note: all implementations of `Unsize` are provided automatically by the compiler, see <https://doc.rust-lang.org/stable/std/marker/trait.Unsize.html> for more information
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
   = note: this error originates in the derive macro `CoercePointee` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0392]: lifetime parameter `'a` is never used
  --> $DIR/deriving-coerce-pointee-neg.rs:25:20
   |
LL | struct NoFieldUnit<'a, #[pointee] T: ?Sized>();
   |                    ^^ unused lifetime parameter
   |
   = help: consider removing `'a`, referring to it in a field, or using a marker such as `PhantomData`

error[E0392]: type parameter `T` is never used
  --> $DIR/deriving-coerce-pointee-neg.rs:25:35
   |
LL | struct NoFieldUnit<'a, #[pointee] T: ?Sized>();
   |                                   ^ unused type parameter
   |
   = help: consider removing `T`, referring to it in a field, or using a marker such as `PhantomData`

error: aborting due to 21 previous errors

Some errors have detailed explanations: E0277, E0392, E0802.
For more information about an error, try `rustc --explain E0277`.
